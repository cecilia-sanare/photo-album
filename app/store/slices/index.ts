import { combineReducers } from '@reduxjs/toolkit';
import photosSlice from './photos.slice';
import albumsSlice from './albums.slice';
import searchSlice from './search.slice';

export const reducers = combineReducers({
  albums: albumsSlice,
  photos: photosSlice,
  search: searchSlice
});
