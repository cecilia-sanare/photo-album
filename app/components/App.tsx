import React, { useEffect, useState } from 'react';
import { useAppDispatch } from '../store/store';
import { Photo, clearPhotos, fetchPhotosByAlbumId, selectPhotos } from '../store/slices/photos.slice';
import { useSelector } from 'react-redux';
import * as styles from './App.module.scss';
import { Header } from './Header';
import { selectSearchQuery, selectSelectedAlbumId, selectSelectedPhoto, setSelectedPhoto } from '../store/slices/search.slice';
import { PhotoCard } from './photos/PhotoCard';
import { PhotoModal } from './photos/PhotoModal';
import { useReadOnlyCachedState } from '../hooks/use-cached-state';

export function App() {
  const dispatch = useAppDispatch();
  const searchQuery = useSelector(selectSearchQuery);
  const selectedAlbumId = useSelector(selectSelectedAlbumId);
  const selectedPhoto = useSelector(selectSelectedPhoto);
  const photos = useSelector(selectPhotos);
  const filteredPhotos = useReadOnlyCachedState(() => {
    if (searchQuery) {
      return photos.filter((photo) => photo.title.includes(searchQuery));
    }

    return photos;
  }, [photos, searchQuery])

  useEffect(() => {
    if (selectedAlbumId) {
      dispatch(fetchPhotosByAlbumId(selectedAlbumId));
    } else {
      dispatch(clearPhotos());
    }
  }, [selectedAlbumId]);

  return (
    <div className={styles.app}>
      <Header />
      <PhotoModal photo={selectedPhoto} />
      <div className={styles.photos}>
        {filteredPhotos.map((photo) => (
          <PhotoCard
            key={photo.id}
            photo={photo}
            onClick={() => dispatch(setSelectedPhoto(photo))}
          />
        ))}
      </div>
    </div>
  );
}
