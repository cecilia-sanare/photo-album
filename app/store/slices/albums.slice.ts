import { createAsyncThunk, createSlice } from '@reduxjs/toolkit'
import { RootState } from '../store';

interface Album {
  userId: number;
  id: number;
  title: string;
}

type AlbumsState = Album[];

const initialState = [] as AlbumsState;

export const fetchAlbums = createAsyncThunk(
  'albums/fetchAlbums',
  async () => {
    const response = await fetch('https://jsonplaceholder.typicode.com/albums');

    return await response.json();
  }
)

const albumsSlice = createSlice({
  name: 'albums',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(fetchAlbums.fulfilled, (state, action) => {
      return action.payload;
    });
  }
});

export const selectAlbums = (state: RootState) => state.albums;

export default albumsSlice.reducer
