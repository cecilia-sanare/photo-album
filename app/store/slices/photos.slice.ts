import { createAsyncThunk, createSelector, createSlice } from '@reduxjs/toolkit'
import type { PayloadAction } from '@reduxjs/toolkit'
import { RootState } from '../store';

export interface Photo {
  albumId: number;
  id: number;
  thumbnailUrl: string;
  title: string;
  url: string;
}

type PhotosState = Photo[];

const initialState = [] as PhotosState;

export const fetchPhotosByAlbumId = createAsyncThunk(
  'photos/fetchByAlbumId',
  async (albumId: number) => {
    const response = await fetch(`https://jsonplaceholder.typicode.com/photos?albumId=${albumId}`);

    return await response.json();
  }
)

const photosSlice = createSlice({
  name: 'photos',
  initialState,
  reducers: {
    setPhotos(state, action: PayloadAction<Photo[]>) {
      return action.payload;
    },
    clearPhotos(state) {
      return [];
    }
  },
  extraReducers: (builder) => {
    builder.addCase(fetchPhotosByAlbumId.fulfilled, (state, action) => {
      return action.payload;
    });
  }
});

export const selectPhotos = (state: RootState) => state.photos;

export const { setPhotos, clearPhotos } = photosSlice.actions
export default photosSlice.reducer
