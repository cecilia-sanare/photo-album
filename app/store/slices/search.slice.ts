import { PayloadAction, createSlice } from '@reduxjs/toolkit'
import { RootState } from '../store';
import { Photo } from './photos.slice';

interface Album {
  userId: number;
  id: number;
  title: string;
}

type searchState = {
  selectedAlbumId?: number;
  query?: string;
  selectedPhoto?: Photo;
};

const initialState = {} as searchState;

const searchSlice = createSlice({
  name: 'search',
  initialState,
  reducers: {
    setSelectedAlbumId(state, action: PayloadAction<number>) {
      return {
        ...state,
        selectedAlbumId: action.payload
      };
    },
    setSearchQuery(state, action: PayloadAction<string>) {
      return {
        ...state,
        query: action.payload
      };
    },
    setSelectedPhoto(state, action: PayloadAction<Photo>) {
      return {
        ...state,
        selectedPhoto: action.payload
      };
    }
  }
});

export const { setSearchQuery, setSelectedAlbumId, setSelectedPhoto } = searchSlice.actions;

export const selectSearch = (state: RootState) => state.search;
export const selectSearchQuery = (state: RootState) => selectSearch(state).query;
export const selectSelectedAlbumId = (state: RootState) => selectSearch(state).selectedAlbumId;
export const selectSelectedPhoto = (state: RootState) => selectSearch(state).selectedPhoto;

export default searchSlice.reducer
