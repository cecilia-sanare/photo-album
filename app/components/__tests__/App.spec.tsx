import { Provider } from 'react-redux';
import { App } from '../App';
import { fireEvent, render } from '@testing-library/react';
import { Chance } from 'chance';
import { configureStore } from '@reduxjs/toolkit';
import { reducers } from '../../store/slices';
import { mockPhoto, mockPhotos } from '../../../__tests__/mocks';

const chance = new Chance();

describe('<App />', () => {
  let store;

  beforeEach(() => {
    store = configureStore({
      reducer: reducers
    });

    jest.spyOn(store, 'dispatch').mockImplementation(() => { });
  });

  it('should render the App', () => {
    const component = render(
      <Provider store={store}>
        <App />
      </Provider>
    );

    expect(component.baseElement).toMatchSnapshot();
  });

  describe('when photos are present', () => {
    it('should render the photos in the given album', () => {
      const albumId = chance.integer();
      const preloadedStore = configureStore({
        preloadedState: {
          photos: mockPhotos(4, {
            albumId: albumId
          }),
          search: {
            selectedAlbumId: albumId
          }
        },
        reducer: reducers
      });

      const component = render(
        <Provider store={preloadedStore}>
          <App />
        </Provider>
      );

      expect(component.getAllByTestId('photo-card')).toHaveLength(4);
    });

    it('should filter the photos by their title', () => {
      const albumId = chance.integer();
      const expectedPhoto = mockPhoto({
        albumId
      });
      const preloadedStore = configureStore({
        preloadedState: {
          photos: [
            ...mockPhotos(4, {
              albumId: albumId
            }),
            expectedPhoto
          ],
          search: {
            selectedAlbumId: albumId,
            query: expectedPhoto.title
          }
        },
        reducer: reducers
      });

      const component = render(
        <Provider store={preloadedStore}>
          <App />
        </Provider>
      );

      expect(component.getAllByTestId('photo-card')).toHaveLength(1);
    });

    describe('and a photo is clicked', () => {
      it('should set the selected photo and open the modal', async () => {
        const albumId = chance.integer();
        const preloadedStore = configureStore({
          preloadedState: {
            photos: [mockPhoto({
              albumId: albumId
            })],
            search: {
              selectedAlbumId: albumId
            }
          },
          reducer: reducers
        });

        const dispatchSpy = jest.spyOn(preloadedStore, 'dispatch');

        const component = render(
          <Provider store={preloadedStore}>
            <App />
          </Provider>
        );

        dispatchSpy.mockClear();

        await fireEvent.click(component.getByTestId('photo-card'));

        expect(preloadedStore.dispatch).toHaveBeenCalledTimes(1);
      });
    });
  });
});
