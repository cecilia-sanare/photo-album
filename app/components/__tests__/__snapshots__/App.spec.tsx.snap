// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render the App 1`] = `
<body>
  <div>
    <div
      class="app"
    >
      <div
        class="header"
      >
        <div
          class="content"
        >
          <input
            class="input search"
            data-testid="input"
            placeholder="Search by the photo name!"
            value=""
          />
          <select
            class="select"
            data-testid="select"
          >
            <option
              disabled=""
              value=""
            >
              Select an Album
            </option>
          </select>
        </div>
      </div>
      <div
        class="backdrop"
        data-testid="backdrop"
      />
      <div
        class="modal"
        data-testid="modal"
      >
        <div
          class="header"
        >
          <div
            class="close"
            data-testid="modal-close"
          >
            x
          </div>
        </div>
        <div
          class="content"
        >
          <img />
        </div>
      </div>
      <div
        class="photos"
      />
    </div>
  </div>
</body>
`;
